src/kv/test/kv_test.cpp

```diff
@@ -1001,3 +1015,110 @@ TEST_CASE("Conflict resolution")
   REQUIRE_THROWS(tx1.commit());
   REQUIRE_THROWS(tx2.commit());
 }
+
+TEST_CASE("Mid-tx compaction")
+{
+  kv::Store kv_store;
+  auto& map_a =
+    kv_store.create<MapTypes::StringNum>("A", kv::SecurityDomain::PUBLIC);
+  auto& map_b =
+    kv_store.create<MapTypes::StringNum>("B", kv::SecurityDomain::PUBLIC);
+
+  constexpr auto key_a = "a";
+  constexpr auto key_b = "b";
+
+  auto increment_vals = [&]() {
+    kv::Tx tx;
+    auto [view_a, view_b] = tx.get_view(map_a, map_b);
+
+    auto a_opt = view_a->get(key_a);
+    auto b_opt = view_b->get(key_b);
+
+    REQUIRE(a_opt == b_opt);
+
+    const auto new_val = a_opt.has_value() ? *a_opt + 1 : 0;
+
+    view_a->put(key_a, new_val);
+    view_b->put(key_b, new_val);
+
+    const auto result = tx.commit();
+    REQUIRE(result == kv::CommitSuccess::OK);
+  };
+
+  increment_vals();
+
+  {
+    INFO("Compaction before get_views");
+    kv::Tx tx;
+
+    increment_vals();
+    kv_store.compact(kv_store.current_version());
+
+    auto view_a = tx.get_view(map_a);
+    auto view_b = tx.get_view(map_b);
+
+    auto a_opt = view_a->get(key_a);
+    auto b_opt = view_b->get(key_b);
+
+    REQUIRE(a_opt == b_opt);
+
+    const auto result = tx.commit();
+    REQUIRE(result == kv::CommitSuccess::OK);
+  }
+
+  {
+    INFO("Compaction after get_views");
+    kv::Tx tx;
+
+    auto view_a = tx.get_view(map_a);
+    increment_vals();
+    auto view_b = tx.get_view(map_b);
+    kv_store.compact(kv_store.current_version());
+
+    auto a_opt = view_a->get(key_a);
+    auto b_opt = view_b->get(key_b);
+
+    REQUIRE(a_opt == b_opt);
+
+    const auto result = tx.commit();
+    REQUIRE(result == kv::CommitSuccess::OK);
+  }
+
+  {
+    INFO("Compaction between get_views");
+    bool threw = false;
+
+    try
+    {
+      kv::Tx tx;
+
+      auto view_a = tx.get_view(map_a);
+      // This transaction does something slow. Meanwhile...
+
+      // ...another transaction commits...
+      increment_vals();
+      // ...and is compacted...
+      kv_store.compact(kv_store.current_version());
+
+      // ...then the original transaction proceeds, expecting to read a single
+      // version
+      auto view_b = tx.get_view(map_b);
```

Comment:
Can we write a comment here or only `try catch` this call instead of the whole block?

Reply:
There's now a comment on this:

```
      // This should throw a CompactedVersionConflict error
```

I prefer try catching the whole block to indicate what the expected behaviour is in _real_ code. We don't want anyone try-catching around individual `get_view`s, we want to catch and retry the entire transaction.